src/Disks/ObjectStorages/Cached/CachedObjectStorage.cpp

```diff
@@ -0,0 +1,300 @@
+#include "CachedObjectStorage.h"
+
+#include <Disks/ObjectStorages/DiskObjectStorageCommon.h>
+#include <IO/BoundedReadBuffer.h>
+#include <Disks/IO/CachedOnDiskWriteBufferFromFile.h>
+#include <Disks/IO/CachedOnDiskReadBufferFromFile.h>
+#include <Common/FileCache.h>
+#include <Common/FileCacheFactory.h>
+#include <Common/CurrentThread.h>
+#include <Common/logger_useful.h>
+#include <filesystem>
+
+namespace fs = std::filesystem;
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int CANNOT_USE_CACHE;
+}
+
+CachedObjectStorage::CachedObjectStorage(
+    ObjectStoragePtr object_storage_,
+    FileCachePtr cache_,
+    const FileCacheSettings & cache_settings_,
+    const std::string & cache_config_name_)
+    : object_storage(object_storage_)
+    , cache(cache_)
+    , cache_settings(cache_settings_)
+    , cache_config_name(cache_config_name_)
+    , log(&Poco::Logger::get(getName()))
+{
+    cache->initialize();
+}
+
+FileCache::Key CachedObjectStorage::getCacheKey(const std::string & path) const
+{
+    return cache->hash(path);
+}
+
+String CachedObjectStorage::getCachePath(const std::string & path) const
+{
+    FileCache::Key cache_key = getCacheKey(path);
+    return cache->getPathInLocalCache(cache_key);
+}
+
+std::string CachedObjectStorage::generateBlobNameForPath(const std::string & path)
+{
+    return object_storage->generateBlobNameForPath(path);
+}
+
+ReadSettings CachedObjectStorage::patchSettings(const ReadSettings & read_settings) const
+{
+    ReadSettings modified_settings{read_settings};
+    modified_settings.remote_fs_cache = cache;
+
+    if (FileCache::isReadOnly())
+        modified_settings.read_from_filesystem_cache_if_exists_otherwise_bypass_cache = true;
+
+    return IObjectStorage::patchSettings(modified_settings);
+}
+
+void CachedObjectStorage::startup()
+{
+    object_storage->startup();
+}
+
+bool CachedObjectStorage::exists(const StoredObject & object) const
+{
+    return object_storage->exists(object);
+}
+
+std::unique_ptr<ReadBufferFromFileBase> CachedObjectStorage::readObjects( /// NOLINT
+    const StoredObjects & objects,
+    const ReadSettings & read_settings,
+    std::optional<size_t> read_hint,
+    std::optional<size_t> file_size) const
+{
+    assert(!objects[0].getPathKeyForCache().empty());
```

Comment:
If `objects` are empty it will fail with segfault before this assertion. In debug, you will get different assertion